subprojects/execution/src/main/java/org/gradle/internal/execution/history/impl/DefaultOverlappingOutputDetector.java

```diff
@@ -50,31 +48,31 @@ public OverlappingOutputs detect(ImmutableSortedMap<String, FileCollectionFinger
         return null;
     }
 
-    private static FileCollectionFingerprint getFingerprintAfterPreviousExecution(@Nullable ImmutableSortedMap<String, FileCollectionFingerprint> previous, String propertyName) {
+    private static FileSystemSnapshot getSnapshotAfterPreviousExecution(@Nullable ImmutableSortedMap<String, FileSystemSnapshot> previous, String propertyName) {
         if (previous != null) {
-            FileCollectionFingerprint afterPreviousExecution = previous.get(propertyName);
+            FileSystemSnapshot afterPreviousExecution = previous.get(propertyName);
             if (afterPreviousExecution != null) {
                 return afterPreviousExecution;
             }
         }
-        return FileCollectionFingerprint.EMPTY;
+        return FileSystemSnapshot.EMPTY;
     }
 
     @Nullable
-    private static OverlappingOutputs detect(String propertyName, FileCollectionFingerprint previous, FileSystemSnapshot before) {
-        Map<String, FileSystemLocationFingerprint> previousFingerprints = previous.getFingerprints();
-        OverlappingOutputsDetectingVisitor outputsDetectingVisitor = new OverlappingOutputsDetectingVisitor(previousFingerprints);
+    private static OverlappingOutputs detect(String propertyName, FileSystemSnapshot previous, FileSystemSnapshot before) {
```

Comment:
We should definitely investigate this direction, as it would lead to significant further simplification of the code. However, it's not a trivial refactor to ensure that we have a single root output snapshot. Let's do that in a separate PR.

Reply:
I added https://github.com/gradle/gradle/issues/15191.

Reply:
Actually, we could check if the root hashes are the same like we do for the up-to-date checks, so we don't need to do a deep visit. That does not depend on us having only one root for the output snapshots.

Reply:
OK, we can do that.